### DISCUSSION TITLE
printf sockaddr_un's sun_path hits BPF stack limit

### DISCUSSION THREAD

**Discussion by martinetd** [bpftrace/bpftrace#2083] [Q&A]
Hello,

I wanted to make a 'connectsnoop' (like opensnoop, but for sys_enter/exit_connect, printing the process and details from the sockaddr; at least address for AF_INET{,6} or unix path)... But I can't seem to figure how to print the unix path, even if I don't do anything else:

```
# bpftrace -e 'tracepoint:syscalls:sys_enter_connect {
  if (args->uservaddr->sa_family == 1 /* AF_UNIX */) {
    printf("%s\n", ((struct sockaddr_un*)args->uservaddr)->sun_path);
  }
}'
error: <unknown>:0:0: in function tracepoint:syscalls:sys_enter_connect i64 (i8*): Looks like the BPF stack limit of 512 bytes is exceeded. Please move large on stack variables into BPF per-cpu array map.
```

Looking at resources about printf allocates a variable on the stack for each element it prints, here sun_path is a `string[108]` so we can see `alloca [108 x i8]` in the assembly dump when running in debug...  But even considering we pay that a couple of times if there are intermediate buffers we're still quite a bit off the 512 max.

Here's the assembly I get:
```
define i64 @"tracepoint:syscalls:sys_enter_connect"(i8* %0) local_unnamed_addr section "s_tracepoint:syscalls:sys_enter_connect_1" {
entry:
  %"struct sockaddr_un.sun_path" = alloca [108 x i8], align 1
  %printf_args = alloca %printf_t, align 8
  %"struct sockaddr.sa_family" = alloca i16, align 2
  %1 = ptrtoint i8* %0 to i64
  %2 = add i64 %1, 24
  %3 = inttoptr i64 %2 to i64*
  %4 = load volatile i64, i64* %3, align 8
  %5 = bitcast i16* %"struct sockaddr.sa_family" to i8*
  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %5)
  %probe_read_user = call i64 inttoptr (i64 112 to i64 (i16*, i32, i64)*)(i16* nonnull %"struct sockaddr.sa_family", i32 2, i64 %4)
  %6 = load i16, i16* %"struct sockaddr.sa_family", align 2
  call void @llvm.lifetime.end.p0i8(i64 -1, i8* nonnull %5)
  %7 = icmp eq i16 %6, 1
  br i1 %7, label %if_body, label %if_end

if_body:                                          ; preds = %entry
  %8 = bitcast %printf_t* %printf_args to i8*
  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %8)
  %9 = bitcast %printf_t* %printf_args to i8*
  call void @llvm.memset.p0i8.i64(i8* nonnull align 8 dereferenceable(120) %9, i8 0, i64 120, i1 false)
  %10 = load volatile i64, i64* %3, align 8
  %11 = add i64 %10, 2
  %12 = getelementptr inbounds [108 x i8], [108 x i8]* %"struct sockaddr_un.sun_path", i64 0, i64 0
  call void @llvm.lifetime.start.p0i8(i64 -1, i8* nonnull %12)
  %probe_read_user1 = call i64 inttoptr (i64 112 to i64 ([108 x i8]*, i32, i64)*)([108 x i8]* nonnull %"struct sockaddr_un.sun_path", i32 108, i64 %11)
  %13 = getelementptr inbounds %printf_t, %printf_t* %printf_args, i64 0, i32 1, i64 0
  call void @llvm.memcpy.p0i8.p0i8.i64(i8* nonnull align 8 dereferenceable(108) %13, i8* nonnull align 1 dereferenceable(108) %12, i64 108, i1 false)
  call void @llvm.lifetime.end.p0i8(i64 -1, i8* nonnull %12)
  %pseudo = call i64 @llvm.bpf.pseudo(i64 1, i64 1)
  %get_cpu_id = call i64 inttoptr (i64 8 to i64 ()*)()
  %perf_event_output = call i64 inttoptr (i64 25 to i64 (i8*, i64, i64, %printf_t*, i64)*)(i8* %0, i64 %pseudo, i64 %get_cpu_id, %printf_t* nonnull %printf_args, i64 120)
  call void @llvm.lifetime.end.p0i8(i64 -1, i8* nonnull %8)
  br label %if_end

if_end:                                           ; preds = %if_body, %entry
  ret i64 0
}
```

So a couple of questions:
 - is that expected? (I'll assume yes?)
 - Would it be possible to print a truncated version of the string?
   - I'd have hoped I could just truncate the string but since it's a buffer str/buf() won't work, and all my attempts at casting failed -- I might have a better chance with the kprobe and $arg1 perhaps... Technically I should also check with addrlen but I didn't even bother with that yet.
 - I can assign it in a map `@addr[tid] = ...->sun_path` but printf does the same with it and print would mean printing the whole map (it refuses `@addr[tid]`) -- OTOH if I'm happy with something not thread-safe as a first approximation assigning to `@addr` works with print. It's not perfect as I can't combine it with the rest of a printf in the same line, but it's better than nothing at all...
 - Any other workaround or solution?

Thanks!

**Comment by martinetd** [comment]
Found https://github.com/iovisor/bpftrace/issues/1147 which sounds like getting a string slice is not possible at the moment, so that answers the second point.
It sounds like I should be able to cast though so I gave it another try, and defining my own struct lookalike of sockaddr_un with a shorter path worked; so I can use that with printf.
That's probably the best I can do right now.

Keeping this open if anyone else has a better idea :)

On a side note, if/when I'm happy with connectsnoop.bt I might post a PR to add it to existing tools as I didn't see any such tool in bpftrace nor bcc

EDIT: ah, string manipulation would still really be great, skipping the first NUL byte is way harder than it should be...

**Reply by mzr** [reply]
Hit the same problem. I tried to apply #1147, but not idea how to cast `char[108]` to `char*`.

```
➜  ~ sudo bpftrace script.bt
script.bt:19:7-14: ERROR: Unknown struct/union: 'char'
      (char *)
      ~~~~~~~
script.bt:18-20: ERROR: str() expects an integer or a pointer type as first argument (none provided)

```

**Reply by mzr** [reply]
Ahh I need to cast to `int8*`

**Reply by mzr** [reply]
It's soooo much easier to define a new struct.

**Reply by xh4n3** [reply]
Hey @mzr , I wonder your solution. IIUC, you are defining your own struct with a `u8 path[108]` field, and then feed into the `str` function, right?

**Reply by mzr** [reply]
I defined a custom structure like this:
```
#include <sys/socket.h>

struct FUNC_PARAMS
{
    __kernel_sa_family_t sun_family;
    char sun_path[20];
};
```

**Reply by xh4n3** [reply]
Edited: The below is doable only if we bypassed the type check in https://github.com/iovisor/bpftrace/blob/v0.16.0/src/ast/passes/semantic_analyser.cpp#L659-L664 , but it needs code change and recompiling. 😢 

---
`char sun_path[108]` will be parse as a String in bpftrace, consuming 108 bytes of the stack memory. When `printf` is called, bpftrace will consume another 108 bytes to construct the data structure passing to `printf`. If we use `u8 sun_path[108]`, bpftrace will parsing this as byte array, the first 108 bytes can be saved by a pointer variable.

So if you write like this, you'll have enough stack memory to get full path, instead of the 20 bytes one.
```
#include <sys/socket.h>

struct FUNC_PARAMS
{
    __kernel_sa_family_t sun_family;
    u8 sun_path[108];
};

tracepoint:syscalls:sys_enter_connect {
  if (args->uservaddr->sa_family == 1 /* AF_UNIX */) {
    $addr = (struct FUNC_PARAMS*)args->uservaddr;
    $path = $addr->sun_path;
    printf("%s\n", str($path, 108));
  }
}
```

And since there was a STRLEN limit, we should run like this:
```bash
BPFTRACE_STRLEN=108 bpftrace long.bt
```

**Reply by martinetd** [reply]
doesn't work for me on bpftrace 0.16.0:

```
# BPFTRACE_STRLEN=108 bpftrace long.bt 
long.bt:13:20-35: ERROR: str() expects an integer or a pointer type as first argument (unsigned int8[108] provided)
    printf("%s\n", str($path, 108));
                   ~~~~~~~~~~~~~~~
```

FWIW, my `connectsnoop.bt` is currently as follow, which works with buffer size 60:
```
#!/usr/bin/bpftrace
/*
 * connectsnoop.bt  Trace connect() syscalls.
 *
 * USAGE: connectsnoop.bt
 *
 * Print connect attempts by tracing the connect() syscall.
 * The address or path is also printed in case of AF_INET,
 * AF_INET6 and AF_UNIX.
 * Failed connects are also displayed.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 *
 * 12-Dec-2021	Dominique Martinet	Created this.
 */

/* redefine sockaddr_un with
 * 1/ a shorter path so we can use it with printf without getting over
 *    the bpf stack limit
 * 2/ the first character of the path split up to be able to test for
 *    abstract sockets (first byte == NUL byte)
 * Note we could just always use the split version and don't really need
 * an union here...
 */
struct sockun {
  short family;
  union {
    char sun_path[60];
    struct {
       char first_char;
       char sun_path_trail[59];
    };
  };
};

BEGIN
{
	printf("Tracing connect syscalls... Hit Ctrl-C to end.\n");
	printf("%-6s %-12s %4s %4s %4s %s\n", "PID", "COMM", "FAMILY", "FD", "ERR", "ADDR");
}


tracepoint:syscalls:sys_enter_connect
{
  @sockets_fd[tid] = args->fd;
  $sock = args->uservaddr;
  @sockets_family[tid] = $sock->sa_family;

  if ($sock->sa_family == 1 /* AF_UNIX */) {
    $sockun = (struct sockun*)$sock;
    if ($sockun->first_char) {
      @sockets_un[tid] = $sockun->sun_path;
    } else {
      @sockets_un_first[tid] = "@";
      @sockets_un[tid] = $sockun->sun_path_trail;
    }
  } else if ($sock->sa_family == 2 /* AF_INET */) {
    /* we need to store this in a different map as the types do not match:
     * ntop returns a value of type 'inet' and we already stored a string */
    @sockets_inet[tid] = ntop(((struct sockaddr_in*)$sock)->sin_addr.s_addr);
  } else if ($sock->sa_family == 10 /* AF_INET6 */) {
    @sockets_inet[tid] = ntop(((struct sockaddr_in6*)$sock)->sin6_addr.in6_u.u6_addr8);
  }
}

tracepoint:syscalls:sys_exit_connect
/@sockets_fd[tid]/
{
	/* unset map entries will print as empty string, leaving only the
	 * desired address */
	printf("%-6d %-16s %2d %4d %4d %s%s%s\n", pid, comm,
			@sockets_family[tid], @sockets_fd[tid], -args->ret,
			@sockets_un_first[tid], @sockets_un[tid],
			@sockets_inet[tid]);
	delete(@sockets_fd[tid]);
	delete(@sockets_family[tid]);
	delete(@sockets_inet[tid]);
	delete(@sockets_un[tid]);
	delete(@sockets_un_first[tid]);
}

END
{
	clear(@sockets_fd);
	clear(@sockets_family);
	clear(@sockets_inet);
	clear(@sockets_un);
	clear(@sockets_un_first);
}
```

no amount of BPFTRACE_STRLEN helps with that if I put in 108, I get told `error: <unknown>:0:0: in function tracepoint:syscalls:sys_exit_connect i64 (i8*): Looks like the BPF stack limit of 512 bytes is exceeded. Please move large on stack variables into BPF per-cpu array map.` as reported in the original message.
The second problem is that if the first byte is a \0 (which is normal for abstract unix sockets) then it is really hard to print the rest of the string, I didn't find anything better than an union and manual checking for it.
